test_quality_of_service/test/test_liveliness.cpp

```diff
@@ -0,0 +1,158 @@
+  subscriber_options.qos_profile = qos_profile;
+
+  // subscriber Liveliness callback event
+  subscriber_options.event_callbacks.liveliness_callback =
+    [&total_number_of_liveliness_events](rclcpp::QOSLivelinessChangedInfo & event) -> void
+    {
+      RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "QOSLivelinessChangedInfo callback");
+      total_number_of_liveliness_events++;
+
+      // strict checking for expected events
+      if (total_number_of_liveliness_events == 1) {
+        // publisher came alive
+        ASSERT_EQ(1, event.alive_count);
+        ASSERT_EQ(0, event.not_alive_count);
+        ASSERT_EQ(1, event.alive_count_change);
+        ASSERT_EQ(0, event.not_alive_count_change);
+      } else if (total_number_of_liveliness_events == 2) {
+        // publisher died
+        ASSERT_EQ(0, event.alive_count);
+        ASSERT_EQ(0, event.not_alive_count);
+        ASSERT_EQ(-1, event.alive_count_change);
+        ASSERT_EQ(0, event.not_alive_count_change);
+      }
+    };
+
+  // publisher options
+  rclcpp::PublisherOptions<> publisher_options;
+  publisher_options.qos_profile = qos_profile;
+
+//  publisher_options.event_callbacks.liveliness_callback =
+//    [](rclcpp::QOSLivelinessLostInfo & event) -> void
+//    {
+//      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "=====QOSLivelinessLostInfo callback fired");
+//    };
+
+  std::string topic("test_automatic_liveliness_changed");
+
+  auto publisher = std::make_shared<Publisher>("publisher", topic, publisher_options, 200ms);
+
+  auto subscriber = std::make_shared<Subscriber>("subscriber", topic, subscriber_options);
+
+  // kill the test after a predetermined amount of time
+  rclcpp::TimerBase::SharedPtr kill_publisher_timer = nullptr;
+  kill_publisher_timer = subscriber->create_wall_timer(
+    2s,
+    [&exec, &publisher, &number_of_published_messages, &kill_publisher_timer]() -> void {
+      exec->remove_node(publisher);
+      number_of_published_messages = publisher->sent_message_count_;
+      publisher.reset();  // force liveliness lost for automatic
+      kill_publisher_timer->cancel();
+    });
+
+  // stop the experiment after a predetermined amount of time
+  rclcpp::TimerBase::SharedPtr kill_experiment_timer = nullptr;
+  kill_experiment_timer = subscriber->create_wall_timer(
+    max_test_length,
+    [&exec, &kill_experiment_timer]() -> void {
+      // don't cancel timer one from here....
+      exec->cancel();
+      kill_experiment_timer->cancel();
+    });
+
+  exec->add_node(subscriber);
+  subscriber->start_listening();
+
+  exec->add_node(publisher);
+  publisher->start_publishing();
+
+  exec->spin();  // blocking
+
+  EXPECT_EQ(2, total_number_of_liveliness_events);  // check expected number of liveliness events
+  EXPECT_GT(number_of_published_messages, 0);  // check if we published anything
+  EXPECT_GT(subscriber->received_message_count_, 0);  // check if we received anything
+}
+
+// todo multiple publisher test?
```

Comment:
Don't leave the todo. I don't think this is necessary for this pass

Reply:
This is stale.